### PR TITLE
Couple of GATK related fixes

### PR DESCRIPTION
**Pull request by mlathara**
Fixes https://github.com/broadinstitute/gatk/issues/7089

Also adds an option to disable remapping of NON_REF for missing allele values.

### CODE REVIEW

File: src/main/cpp/src/utils/vid_mapper_pb.cc
Comment (nalinigans) [inline src/main/cpp/src/utils/vid_mapper_pb.cc:373]: Wondering why tests/golden_outputs/t0_1_2_all_asa_java_query_vcf_no_remap_missing does not exercise this code or is it a codecov quirk? Otherwise is it easy enough to write a test  to cover this logic?

**Review by nalinigans — COMMENTED** [review]
Can you wait until #144 is merged to get the MacOS builds to run?

File: src/main/cpp/src/utils/vid_mapper_pb.cc
Comment (mlathara) [reply src/main/cpp/src/utils/vid_mapper_pb.cc:373]: The CI test uses the query json while this is protobuf. I'll add a test for this

**Comment by mlathara** [comment]
> Can you wait until #144 is merged to get the MacOS builds to run?

Sure

File: src/main/cpp/include/config/genomicsdb_config_base.h
Comment (kgururaj) [inline src/main/cpp/include/config/genomicsdb_config_base.h:166]: Does this need to be a configuration variable? I'm ok with simply trusting the input/reference genome

File: src/main/cpp/src/query_operations/variant_operations.cc
Comment (kgururaj) [inline src/main/cpp/src/query_operations/variant_operations.cc:464]: Can we generalize this to other fields also? Not just allele specific annotations

File: src/main/cpp/src/utils/vid_mapper.cc
Comment (kgururaj) [inline src/main/cpp/src/utils/vid_mapper.cc:1308]: Move the isBool condition to a VERIFY_OR_THROW

**Review by kgururaj — COMMENTED** [review]
I believe there are a few [allele specific annotation fields in the Java side](https://github.com/GenomicsDB/GenomicsDB/blob/master/src/main/java/org/genomicsdb/importer/Constants.java) which are hard coded into the Protobuf vid - can you check if the disable remap needs to be set on the Java side for these fields?

**Comment by nalinigans** [comment]
> > Can you wait until #144 is merged to get the MacOS builds to run?
> 
> Sure

#144 is merged into develop!

File: src/main/cpp/include/config/genomicsdb_config_base.h
Comment (mlathara) [reply src/main/cpp/include/config/genomicsdb_config_base.h:166]: Yeah I wasn't sure if we wanted the config knob either...from the sounds of it, GATK just capitalizes the bases. 

When we asked folks about whether any analyses might care about the distinction that the base was soft masked per the reference we just got a 🤷‍♂️ 

I'm fine with removing the config knob -- but I think our behavior should be to capitalize the base to stay consistent with GATK.

File: src/main/cpp/src/query_operations/variant_operations.cc
Comment (mlathara) [reply src/main/cpp/src/query_operations/variant_operations.cc:464]: I started doing this, but had a question when doing some tests for these. Info fields whose length is based on number of alleles (like AD) are sorta easy to reason about with this remapping. But what's the correct behavior for things like PL with length based on number of possible genotypes? Any genotype that includes the missing allele should be missing if the remap is enabled?

Also, it seems like remap_GT_fields also does something like this...but I'm not sure I understand the case where that is required. Won't the merged set of alleles always have the alleles required by each of of the sample's genotypes?

**Comment by mlathara** [comment]
- Removed the knob for soft masking. Now we capitalize the reference by default - is that okay with you all?
- Added the more general disable remap for other fields...but not sure about the test for it because I don't know of any other tool to get the golden output. They make sense to me but might have missed something
- I didn't hardcode any fields as disable remap -- figure we can leave this for the GATK PR rather than have GenomicsDB worry about it.

File: src/main/cpp/include/config/genomicsdb_config_base.h
Comment (kgururaj) [reply src/main/cpp/include/config/genomicsdb_config_base.h]: Ah ok makes sense

File: src/main/cpp/src/query_operations/variant_operations.cc
Comment (kgururaj) [reply src/main/cpp/src/query_operations/variant_operations.cc]: > I started doing this, but had a question when doing some tests for these. Info fields whose length is based on number of alleles (like AD) are sorta easy to reason about with this remapping. But what's the correct behavior for things like PL with length based on number of possible genotypes? Any genotype that includes the missing allele should be missing if the remap is enabled?
Yes
> 
> Also, it seems like remap_GT_fields also does something like this...but I'm not sure I understand the case where that is required. Won't the merged set of alleles always have the alleles required by each of of the sample's genotypes?
Right,  GT shouldn't have this issue

**Comment by kgururaj** [comment]
> - Removed the knob for soft masking. Now we capitalize the reference by default - is that okay with you all?
> - Added the more general disable remap for other fields...but not sure about the test for it because I don't know of any other tool to get the golden output. They make sense to me but might have missed something
> - I didn't hardcode any fields as disable remap -- figure we can leave this for the GATK PR rather than have GenomicsDB worry about it.

Can you check if the 3 fields are fixed by GATK and they're not relying on us for the combination?

**Review by kgururaj — APPROVED** [review]

**Review by nalinigans — APPROVED** [review]
Looks good to me too. I will start a snapshot build as soon as you merge this in. Thanks

**Comment by bbimber** [comment]
@mlathara Thanks for working on this. I just saw this change come though. Just so I understand the implications:

- Is there a corresponding switch on the GATK side to toggle this behavior? 
- what is the default behavior with respect to soft-clipping when running GenotypeGVCFs for the latest GATK version at this point?

Just so I understand, if we have a pre-existing genomics DB workspace created using an older version of GATK GenomicsDbImport, are the soft-clipped reference bases stored in that workspace such that we need to remake it? Or if I re-run GenotypeGVCFs on this pre-existing workspace using the newest GATK will it report the true REF even over soft-clipped sites?

**Comment by mlathara** [comment]
@bbimber This hasn't yet made it into master -- so we don't have a release that's been incorporated into GATK yet. Working towards that -- hopefully we'll have a PR for that in the next week or so. To answer your questions:
- There will be no switch on the GATK side to toggle this behavior
- The latest GATK doesn't have these changes yet, so nothing has changed yet. Once this is incorporated there, any soft masked regions will show up with the true REF values not `N` (the bases will be capitalized, regardless of what the reference indicates)
- Once this release is incorporated into GATK, querying GenomicsDB workspaces created with a version of GenomicsDBImport prior to this fix will return the true REF values not `N` (capitalized bases as mentioned above)

**Comment by bbimber** [comment]
@mlathara OK, got it. That third point is great news. I am very happy not to need to remake our workspaces.

**Comment by bbimber** [comment]
Hello - just thought I'd check back. did you have any better idea when the fixed might go into GenomicsDB and also the PR for GATK? If there's anything we can do to help please let me know.

**Comment by mlathara** [comment]
The GATK PR is actually already up: https://github.com/broadinstitute/gatk/pull/7224 I'm gonna put up a separate PR with a GATK test for this fix -- but you'll have the fix once that PR gets accepted.

**Pull request merged**
